flytekit/experimental/eager_function.py

```diff
@@ -0,0 +1,569 @@
+            entity_type=node.entity_type,
+            entity_name=node.entity_name,
+            execution_name=node.execution.id.name,
+            url=node.url,
+            inputs=node_inputs,
+            outputs=node_outputs,
+        )
+
+    Deck("eager workflow", output)
+
+
+@asynccontextmanager
+async def eager_context(
+    fn,
+    remote: FlyteRemote,
+    ctx: FlyteContext,
+    async_stack: AsyncStack,
+):
+    """This context manager overrides all tasks in the global namespace with async versions."""
+
+    _original_cache = {}
+
+    # override tasks with async version
+    for k, v in fn.__globals__.items():
+        if isinstance(v, (PythonTask, WorkflowBase)):
+            _original_cache[k] = v
+            fn.__globals__[k] = AsyncEntity(v, remote, ctx, async_stack)
+
+    try:
+        yield
+    finally:
+        # restore old tasks
+        for k, v in _original_cache.items():
+            fn.__globals__[k] = v
+
+
+async def node_cleanup_async(sig, loop, async_stack: AsyncStack):
+    """Clean up subtasks when eager workflow parent is done.
+
+    This applies either if the eager workflow completes successfully, fails, or is cancelled by the user.
+    """
+    logger.debug(f"Cleaning up async nodes on signal: {sig}")
+    terminations = []
+    for node in async_stack.call_stack:
+        terminations.append(node.async_entity.terminate())
+    results = await asyncio.gather(*terminations)
+    logger.debug(f"Successfully terminated subtasks {results}")
+
+
+def node_cleanup(sig, frame, loop, async_stack: AsyncStack):
+    """Clean up subtasks when eager workflow parent is done.
+
+    This applies either if the eager workflow completes successfully, fails, or is cancelled by the user.
+    """
+    logger.debug(f"Cleaning up async nodes on signal: {sig}")
+    terminations = []
+    for node in async_stack.call_stack:
+        terminations.append(node.async_entity.terminate())
+    results = asyncio.gather(*terminations)
+    results = asyncio.run(results)
+    logger.debug(f"Successfully terminated subtasks {results}")
+    loop.close()
+
+
+def eager(
+    _fn=None,
+    *,
+    remote: Optional[FlyteRemote] = None,
+    client_secret_group: Optional[str] = None,
+    client_secret_key: Optional[str] = None,
+    **kwargs,
+):
+    """Eager workflow decorator.
+
+    This type of workflow will execute all flyte entities within it eagerly, meaning that all python constructs can be
+    used inside of an ``@eager``-decoratod function. This is because eager workflows use a
+    :py:class:`~flytekit.remote.remote.FlyteRemote` object to kick off executions when a flyte entity needs to produce a
+    value.
+
+    For example:
+
+    .. code-block:: python
+
+        from flytekit import task
+        from flytekit.experimental import eager
+
+        @task
+        def add_one(x: int) -> int:
+            return x + 1
+
+        @task
+        def double(x: int) -> int:
+            return x * 2
+
+        @eager
```

Comment:
qq: are we able to pass image or image spec here?

Reply:
yep! you can forward any kwarg that `@task` accepts